osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,585 @@
+                                                              "/etc/rc4.d",
+                                                              "/etc/rc5.d",
+                                                              "/etc/rc6.d",
+                                                              "/etc/rcS.d"};
+
+  std::unordered_map<std::string, std::vector<size_t>> service_list;
+
+  for (size_t run_level = 0U; run_level < run_level_dir_list.size();
+       ++run_level) {
+    const auto& run_level_path = run_level_dir_list[run_level];
+
+    try {
+      boostfs::directory_iterator end_it;
+      for (auto it = boostfs::directory_iterator(run_level_path); it != end_it;
+           ++it) {
+        const auto& symlink = it->path();
+        if (!boostfs::is_symlink(symlink)) {
+          continue;
+        }
+
+        auto symlink_destination = boostfs::canonical(symlink);
+        if (!boostfs::is_regular_file(symlink_destination)) {
+          VLOG(1) << "Skipping broken symlink: " << symlink << " -> "
+                  << symlink_destination;
+
+          continue;
+        }
+
+        auto service_name = symlink_destination.filename().string();
+        service_list[service_name].push_back(run_level);
+      }
+    } catch (const boostfs::filesystem_error& e) {
+      VLOG(1) << "An error has occurred: " << e.what() << ". Continuing anyway";
+    }
+  }
+
+  for (const auto& p : service_list) {
+    const auto& service_name = p.first;
+    const auto& run_level_list = p.second;
+
+    Row r;
+    r["name"] = service_name;
+    r["status"] = "";
+    r["path"] = std::string("/etc/init.d/") + service_name;
+    r["service_type"] = "SysVinit";
+
+    // Use the same syntax as Upstart
+    if (!run_level_list.empty()) {
+      r["start_type"] = "runlevel [";
+
+      for (const auto& run_level : run_level_list) {
+        if (run_level != 7) {
+          r["start_type"] += std::to_string(run_level);
+        } else {
+          r["start_type"] += "S";
+        }
+      }
+
+      r["start_type"] += "]";
+    }
+
+    query_data.push_back(r);
+  }
+
+  return Status(0, "OK");
+}
+
+sd_bus* getSystemdBusHandle() {
+  /*
+    https://github.com/systemd/systemd/blob/master/ENVIRONMENT.md
+
+    $SYSTEMCTL_FORCE_BUS=1
+      if set, do not connect to PID1's private D-Bus listener, and
+      instead always connect through the dbus-daemon D-bus broker.
+  */
+
+  auto force_dbus_env_var = std::getenv("SYSTEMCTL_FORCE_BUS");
+
+  bool connection_mode = true;
+  if (force_dbus_env_var != nullptr) {
+    connection_mode = boost::lexical_cast<bool>(force_dbus_env_var);
+  }
+
+  if (connection_mode || geteuid() != 0) {
+    sd_bus* bus = nullptr;
+    if (sd_bus_default_system(&bus) < 0) {
+      return nullptr;
+    }
+
+    sd_bus_set_allow_interactive_authorization(bus, 0);
+    return bus;
+  }
+
+  sd_bus* bus = nullptr;
+
+  try {
+    if (sd_bus_new(&bus) < 0 || bus == nullptr) {
+      throw std::runtime_error("Failed to start the bus");
```

Comment:
If you want to use a throwing-API can you create a specific exception for this function and catch it specifically where it's used?

Reply:
I've removed all exceptions from the code